lib/Sema/TypeCheckAvailability.cpp

```diff
@@ -1499,6 +1499,27 @@ bool TypeChecker::isInsideUnavailableDeclaration(
                                   IsUnavailable);
 }
 
+bool TypeChecker::isInsideCompatibleUnavailableDeclaration(
+	SourceRange ReferenceRange, const DeclContext *ReferenceDC,
+	const AvailableAttr *attr)
+{
+	if (!attr->isUnconditionallyUnavailable()) {
+		return false;
+	}
+	PlatformKind platform = attr->Platform;
+	if (platform == PlatformKind::none) {
+		return false;
+	}
+
+    auto IsUnavailable = [platform](const Decl *D) {
+      auto EnclosingUnavailable = D->getAttrs().getUnavailable(D->getASTContext());
+	  return EnclosingUnavailable && EnclosingUnavailable->Platform == platform;
+    };
+
+    return someEnclosingDeclMatches(ReferenceRange, ReferenceDC, *this,
```

Comment:
Hm, this doesn't quite seem like the right query. If a declaration is unavailable on every platform, and then the *next* enclosing declaration is unavailable in a compatible way, this will return `true` rather than `false`.

Reply:
I'm not sure I understand, but here goes ... my understanding is that the `Platform` of the `AvailableAttr` is always either `::none` or the *current compiling* platform by this point. It seems like `AvailableAttr` holds the sum of the reasons the current declaration is unavailable in the current compile (introduction/deprecation versions, platform), but it does not hold a complete record of every `@available` annotation in the source.

So, when I try to add a test that sounds like what you are suggesting, it works as I would hope:

	// RUN: %target-typecheck-verify-swift
	// REQUIRES: OS=macosx

	// Check that only the platform we are compiling for is considered when allowing access to unavailable declarations.

	@available(iOS, unavailable)
	@available(OSX, unavailable)
	func unavailable() {} // expected-note {{'unavailable()' has been explicitly marked unavailable here}}

	@available(iOS, unavailable)
	func call_with_ios_unavailable() {
		unavailable() // expected-error {{'unavailable()' is unavailable}}
	}

	@available(OSX, unavailable)
	func call_with_osx_unavailable() {
		unavailable() // OK: same
	}

Adding debugging code to `TypeChecker::isInsideCompatibleUnavailableDeclaration()`, it is called twice. In both cases the `Platform` is 4 (`OSX`). That is, the `@available(iOS, unavailable)` enclosing the first call site has been discarded by this point since we aren't compiling for iOS, and it doesn't permit the call to `unavailable()` from `call_with_ios_unavailable()`.

Reply:
Sorry, I should have just included a code example.

```swift
// on macOS
@available(iOS, unavailable)
class Outer {
  @available(*, unavailable)
  func completelyBadMethod() {}
}

@available(iOS, unavailable)
func test(outer: Outer) {
  outer.completelyBadMethod() // should still error
}
```

On top of that, your example is also unfortunate, because it means the active and non-active platforms don't behave the same. But I think that's inevitable, since you can always imagine a hypothetical dishwasherOS platform that can call everything.

Reply:
Ah, got it. Test added and passes locally. Also added a swapped version of the nested availability, which fails to build as expected since the outer declaration can't be referenced.

Reply:
Ah, hm. Now I need to figure out why this works. I really thought it would identify the outer context as being compatible and then return true.

Reply:
I believe it is that the 'completely unavailable' declaration passes the first check:

	if (!attr->isUnconditionallyUnavailable()) {
		return false;
	}

but fails the second:

	PlatformKind platform = attr->Platform;
	if (platform == PlatformKind::none) {
		return false;
	}

and so we don't look at any enclosing source ranges.

Reply:
It seems possible that a more complicated stack of nested availabilities might provoke the issue you were worried about. I'll can poke at it a bit more after my current `--test` build finishes.

Reply:
So far I haven't been able to write a test that breaks. Any stack of iOS/unavailable/iOS stops at the point the caller tries to spell the unavailable symbols and doesn't look at the child symbols. And the test you suggested hits the path mentioned above.

Perhaps I should add some commentary to make the intention more clear.

Reply:
Not sure sure if this is more clear for future readers or not... thoughts?


	// If a declaration has an unavailable attribute applied, it can still
	// called if some enclosing scope of the call site has the same unavailability.
	bool TypeChecker::isInsideCompatibleUnavailableDeclaration(
	    SourceRange ReferenceRange, const DeclContext *ReferenceDC,
	    const AvailableAttr *attr) {
	  // We aren't attempting to do introduced/deprecated version checks
	  if (!attr->isUnconditionallyUnavailable()) {
	    return false;
	  }
  
	  // Only allow per-platform unavailability for the current platform.
	  // At this point, the platform will either be ::none (totally unavailable)
	  // and not allowed here, or the current platform. All other annotations
	  // of `@unavailable(OtherPlatform, *)` will have been dropped already.
	  PlatformKind platform = attr->Platform;
	  if (platform == PlatformKind::none) {
	    return false;
	  }

	  auto IsUnavailable = [platform](const Decl *D) {
	    auto EnclosingUnavailable =
	        D->getAttrs().getUnavailable(D->getASTContext());
	    return EnclosingUnavailable && EnclosingUnavailable->Platform == platform;
	  };

	  return someEnclosingDeclMatches(ReferenceRange, ReferenceDC, *this,
	                                  IsUnavailable);
	}